parseutils.c

```diff
@@ -355,69 +358,178 @@ parse_ss_line(char *line, int *ntok)
 }
 
 /*
- * strip_quotes (lifted and modifed from src/bin/psql/stringutils)
+ * parse_quoted_string
  *
- * Remove quotes from the string at *source.  Leading and trailing occurrences
- * of 'quote' are removed.
+ * Remove quotes and escapes from the string at **source.  Returns a new palloc() string with the contents.
  *
- * Note that the source string is overwritten in-place.
  */
-void
-strip_quotes(char *source, char quote)
+char*
+parse_quoted_string(char **source)
 {
 	char	   *src;
 	char	   *dst;
+	char	   *ret;
+	bool        lastSlash = false;
 
 	Assert(source != NULL);
-	Assert(quote != '\0');
+	Assert(*source != NULL);
 
-	src = dst = source;
+	src = *source;
+	ret = dst = palloc0(strlen(src));
 
-	if (*src && *src == quote)
+	if (*src && *src == '"')
 		src++;					/* skip leading quote */
 
 	while (*src)
 	{
 		char		c = *src;
+		pg_wchar    cp = 0;
+
+		if (lastSlash)
+		{
+			switch (c) {
+			case '\\':
+				*dst++ = '\\';
+				src++;
+				break;
+			case 'a':
+				*dst++ = '\a';
+				src++;
+				break;
+			case 'b':
+				*dst++ = '\b';
+				src++;
+				break;
+			case 'f':
+				*dst++ = '\f';
+				src++;
+				break;
+			case 'n':
+				*dst++ = '\n';
+				src++;
+				break;
+			case 'r':
+				*dst++ = '\r';
+				src++;
+				break;
+			case 't':
+				*dst++ = '\t';
+				src++;
+				break;
+			case 'v':
+				*dst++ = '\v';
+				src++;
+				break;
+			case '"':
+				*dst++ = '"';
+				src++;
+				break;
+			case 'x':
+				/* next 2 chars are hex bytes */
+				if (is_hex_digit(src[1]) && is_hex_digit(src[2]))
```

Comment:
Should this (and the 'u' 'U' loop below) be more defensive about the length of src? The last two characters the buffer being `\x`, for example.

Reply:
The `is_hex_digit()` check catches those cases; a nul byte fails the check,
so it'll hit the error path and bail.